File: .agents/skills/unqa-swarm/scripts/provision.py
#!/usr/bin/env python3
"""
UNQA Swarm — Workspace Provisioning Script (Python)

Parses a masterplan.md DAG table and generates:
  - router.json (lightweight routing table)
  - Task folders with task.json and instruction.md per task
  - done/ directory

Usage: python3 provision.py --job-dir .unqa-swarm_workspace/job_[id]/
"""

import json
import os
import re
import sys
from datetime import datetime, timezone
from pathlib import Path


def parse_array(s: str) -> list:
    """Parse array-like strings: ['item1', 'item2'] or plain text."""
    cleaned = s.replace('`', '').strip()
    if cleaned.startswith('['):
        try:
            return json.loads(cleaned.replace("'", '"'))
        except json.JSONDecodeError:
            cleaned = re.sub(r"[\[\]']", '', cleaned)
            return [x.strip() for x in cleaned.split(',') if x.strip()]
    if cleaned in ('None', '-', ''):
        return []
    return [x.strip() for x in cleaned.split(',') if x.strip()]


def main():
    # --- Argument Parsing ---
    if '--job-dir' not in sys.argv:
        print(json.dumps({'success': False, 'error': 'Missing --job-dir argument'}))
        sys.exit(1)

    job_dir_idx = sys.argv.index('--job-dir')
    if job_dir_idx + 1 >= len(sys.argv):
        print(json.dumps({'success': False, 'error': 'Missing --job-dir value'}))
        sys.exit(1)

    job_dir = Path(sys.argv[job_dir_idx + 1]).resolve()

    # --- Read Masterplan ---
    masterplan_path = job_dir / 'masterplan.md'
    if not masterplan_path.exists():
        print(json.dumps({'success': False, 'error': f'masterplan.md not found at {masterplan_path}'}))
        sys.exit(1)

    masterplan = masterplan_path.read_text(encoding='utf-8')

    # --- Extract Metadata ---
    job_id = job_dir.name
    project_match = re.search(r'^#\s+.*?Masterplan:\s*(.+)$', masterplan, re.MULTILINE)
    project_name = project_match.group(1).strip() if project_match else job_id
    architect_match = re.search(r'\*\*Architect:\*\*\s*(.+)$', masterplan, re.MULTILINE)
    architect_name = architect_match.group(1).strip() if architect_match else 'Unknown'

    # --- Parse DAG Table ---
    lines = masterplan.split('\n')
    table_start: int = -1
    for i, line in enumerate(lines):
        if re.match(r'^\|\s*(Task|Chunk)\s*ID', line, re.IGNORECASE):
            table_start = i
            break

    if table_start == -1:
        print(json.dumps({'success': False, 'error': 'No DAG table found in masterplan.md'}))
        sys.exit(1)

    # Parse headers dynamically
    header_line = lines[table_start]
    headers = [h.strip().lower() for h in header_line.split('|') if h.strip()]
    
    col_map = {
        'taskId': next((i for i, h in enumerate(headers) if 'id' in h), -1),
        'targetFile': next((i for i, h in enumerate(headers) if 'target' in h or 'file' in h), -1),
        'dependencies': next((i for i, h in enumerate(headers) if 'dep' in h), -1),
        'skills': next((i for i, h in enumerate(headers) if 'skill' in h), -1),
        'contextFiles': next((i for i, h in enumerate(headers) if 'context' in h or 'read' in h), -1),
        'antiGoals': next((i for i, h in enumerate(headers) if 'anti' in h), -1),
        'description': next((i for i, h in enumerate(headers) if 'desc' in h), -1)
    }

    # Fallback missing cols
    if col_map['taskId'] == -1: col_map['taskId'] = 0
    if col_map['targetFile'] == -1: col_map['targetFile'] = 1
    if col_map['dependencies'] == -1: col_map['dependencies'] = 2
    if col_map['skills'] == -1: col_map['skills'] = 3
    if col_map['description'] == -1: col_map['description'] = len(headers) - 1

    tasks = []
    start_index = int(table_start) + 2
    for i in range(start_index, len(lines)):
        line = lines[i].strip()
        if not line.startswith('|'):
            break

        # A simple split handles most markdown table rows safely
        raw_cols = line.split('|')
        if len(raw_cols) > 0 and raw_cols[0].strip() == '': raw_cols.pop(0)
        if len(raw_cols) > 0 and raw_cols[-1].strip() == '': raw_cols.pop(-1)
        
        cols = [c.strip() for c in raw_cols]
        if len(cols) < 2:
            continue

        def extract_col(idx):
            if idx != -1 and idx < len(cols):
                return cols[idx]
            return ''

        # Sanitize task ID
        raw_task_id = extract_col(col_map['taskId']).replace('`', '').strip()
        task_id = re.sub(r'[^a-zA-Z0-9_\-]', '_', raw_task_id)
        if not task_id:
            continue

        target_file = extract_col(col_map['targetFile']).replace('`', '').strip()

        # Dependencies
        deps_raw = extract_col(col_map['dependencies'])
        if deps_raw in ('None', '-', ''):
            dependencies = []
        else:
            dependencies = [d.strip() for d in deps_raw.replace('`', '').split(',') if d.strip()]

        assigned_skills = parse_array(extract_col(col_map['skills']))
        context_files = parse_array(extract_col(col_map['contextFiles']))
        anti_goals = parse_array(extract_col(col_map['antiGoals']))
        description = extract_col(col_map['description'])

        tasks.append({
            'task_id': task_id,
            'description': description,
            'dependencies': dependencies,
            'assigned_skills': assigned_skills,
            'target_file': target_file,
            'context_files_to_read': context_files,
            'anti_goals': anti_goals,
        })

    if not tasks:
        print(json.dumps({'success': False, 'error': 'No tasks parsed from DAG table'}))
        sys.exit(1)

    # --- Parse API Contracts ---
    api_contracts = {}
    contract_match = re.search(r'##\s*🔌\s*API\s*Contracts([\s\S]*?)(?=\n##\s|$)', masterplan)
    if contract_match:
        json_blocks = re.findall(r'```json\n([\s\S]*?)```', contract_match.group(1))
        for block in json_blocks:
            try:
                parsed = json.loads(block)
                api_contracts.update(parsed)
            except json.JSONDecodeError:
                pass

    # --- Generate router.json ---
    now = datetime.now(timezone.utc).isoformat()
    task_statuses = {t['task_id']: 'pending' for t in tasks}

    router = {
        'job_id': job_id,
        'project_name': project_name,
        'architect_name': architect_name,
        'manager_name': 'provision-script',
        'owner_name': None,
        'global_status': 'executing',
        'batch_limit': 3,
        'created_at': now,
        'updated_at': now,
        'tasks': task_statuses,
    }

    router_path = job_dir / 'router.json'
    router_path.write_text(json.dumps(router, indent=2), encoding='utf-8')

    # --- Create done/ directory ---
    done_dir = job_dir / 'done'
    done_dir.mkdir(parents=True, exist_ok=True)

    # --- Read Templates ---
    skill_root = Path(__file__).resolve().parent.parent
    task_schema_path = skill_root / 'templates' / 'task_schema.json'
    instruction_template_path = skill_root / 'templates' / 'instruction_template.md'

    instruction_template = (
        instruction_template_path.read_text(encoding='utf-8')
        if instruction_template_path.exists() else None
    )

    # --- Provision Each Task ---
    created_folders = []

    for task in tasks:
        task_id: str = str(task['task_id'])
        task_dir = job_dir / task_id
        task_dir.mkdir(parents=True, exist_ok=True)

        # Generate task.json
        task_json = {
            'task_id': task_id,
            'description': task['description'],
            'dependencies': task['dependencies'],
            'assigned_skills': task['assigned_skills'],
            'target_file': task['target_file'],
            'current_status': 'pending',
            'workers_assigned': [],
            'context_files_to_read': task['context_files_to_read'],
            'anti_goals': task['anti_goals'],
            'api_contract': api_contracts,
            'status_history': [
                {
                    'status': 'pending',
                    'timestamp': now,
                    'changed_by': 'provision-script',
                    'notes': 'Mapped from masterplan.',
                }
            ],
        }

        (task_dir / 'task.json').write_text(
            json.dumps(task_json, indent=2), encoding='utf-8'
        )

        # Generate instruction.md
        if instruction_template is not None:
            instruction = str(instruction_template)
            instruction = instruction.replace('{{task_id}}', task_id)
            instruction = instruction.replace('{{target_file_path}}', str(task['target_file']))
            instruction = instruction.replace(
                '{{assigned_skills_array}}', json.dumps(task['assigned_skills'])
            )
            
            deps = task['dependencies']
            if isinstance(deps, list) and deps:
                deps_str = ', '.join(str(d) for d in deps)
            else:
                deps_str = 'None'
            
            instruction = instruction.replace('{{dependencies_list_or_none}}', deps_str)

            ctx_list = '\n'.join(f'* `{f}`' for f in task['context_files_to_read']) if isinstance(task['context_files_to_read'], list) else ''
            instruction = re.sub(r'\{\{context_files_to_read_list.*?\}\}', ctx_list, instruction)

            ag_list = '\n'.join(f'* {g}' for g in task['anti_goals']) if isinstance(task['anti_goals'], list) else ''
            instruction = re.sub(r'\{\{anti_goals_list.*?\}\}', ag_list, instruction)

            instruction = instruction.replace(
                '{{api_contract_from_task_json}}',
                json.dumps(api_contracts, indent=2),
            )

            instruction = re.sub(
                r'\{\{Specific description.*?\}\}', str(task['description']), instruction
            )

            (task_dir / 'instruction.md').write_text(instruction, encoding='utf-8')

        created_folders.append(task_id)

    # --- Output Result ---
    print(json.dumps({
        'success': True,
        'job_id': job_id,
        'tasks_provisioned': len(created_folders),
        'task_ids': created_folders,
        'router_path': str(router_path),
        'done_dir': str(done_dir),
    }))


if __name__ == '__main__':
    main()

File: .agents/skills/unqa-swarm/scripts/provision.js
#!/usr/bin/env node
/**
 * UNQA Swarm — Workspace Provisioning Script (Node.js)
 * 
 * Parses a masterplan.md DAG table and generates:
 *   - router.json (lightweight routing table)
 *   - Task folders with task.json and instruction.md per task
 *   - done/ directory
 * 
 * Usage: node provision.js --job-dir .unqa-swarm_workspace/job_[id]/
 */

const fs = require('fs');
const path = require('path');

// --- Argument Parsing ---
const args = process.argv.slice(2);
const jobDirIndex = args.indexOf('--job-dir');
if (jobDirIndex === -1 || !args[jobDirIndex + 1]) {
  console.error(JSON.stringify({ success: false, error: 'Missing --job-dir argument' }));
  process.exit(1);
}
const jobDir = path.resolve(args[jobDirIndex + 1]);

// --- Read Masterplan ---
const masterplanPath = path.join(jobDir, 'masterplan.md');
if (!fs.existsSync(masterplanPath)) {
  console.error(JSON.stringify({ success: false, error: `masterplan.md not found at ${masterplanPath}` }));
  process.exit(1);
}
const masterplan = fs.readFileSync(masterplanPath, 'utf-8');

// --- Extract Metadata ---
const jobId = path.basename(jobDir);
const projectNameMatch = masterplan.match(/^#\s+.*?Masterplan:\s*(.+)$/m);
const projectName = projectNameMatch ? projectNameMatch[1].trim() : jobId;
const architectMatch = masterplan.match(/\*\*Architect:\*\*\s*(.+)$/m);
const architectName = architectMatch ? architectMatch[1].trim() : 'Unknown';

// --- Parse DAG Table ---
const lines = masterplan.split('\n');
const tableStart = lines.findIndex(l => /^\|\s*(Task|Chunk)\s*ID/i.test(l));
if (tableStart === -1) {
  console.error(JSON.stringify({ success: false, error: 'No DAG table found in masterplan.md' }));
  process.exit(1);
}

// Parse headers dynamically
const headers = lines[tableStart].split('|').map(h => h.trim().toLowerCase()).filter(h => h.length > 0);

const colMap = {
  taskId: headers.findIndex(h => h.includes('id')),
  targetFile: headers.findIndex(h => h.includes('target') || h.includes('file')),
  dependencies: headers.findIndex(h => h.includes('dep')),
  skills: headers.findIndex(h => h.includes('skill')),
  contextFiles: headers.findIndex(h => h.includes('context') || h.includes('read')),
  antiGoals: headers.findIndex(h => h.includes('anti')),
  description: headers.findIndex(h => h.includes('desc'))
};

// Fallback for essential missing columns based on template defaults
if (colMap.taskId === -1) colMap.taskId = 0;
if (colMap.targetFile === -1) colMap.targetFile = 1;
if (colMap.dependencies === -1) colMap.dependencies = 2;
if (colMap.skills === -1) colMap.skills = 3;
if (colMap.description === -1) colMap.description = headers.length - 1;

const tasks = [];
for (let i = tableStart + 2; i < lines.length; i++) {
  const line = lines[i].trim();
  if (!line.startsWith('|')) break;
  
  // We use a regex split to avoid breaking on escaped pipes, though simple LLM tables rarely use them
  const rawCols = line.split('|');
  // the first and last split elements are usually empty because of leading/trailing pipes
  rawCols.shift();
  if (rawCols.length > 0 && rawCols[rawCols.length - 1].trim() === '') rawCols.pop();
  const cols = rawCols.map(c => c.trim());
  
  if (cols.length < 2) continue; // Need at least ID and Target
  
  const extractCol = (idx) => (idx !== -1 && idx < cols.length) ? cols[idx] : '';
  
  // Sanitize task ID to prevent path traversal
  const rawTaskId = extractCol(colMap.taskId).replace(/`/g, '').trim();
  const taskId = rawTaskId.replace(/[^a-zA-Z0-9_-]/g, '_');
  if (!taskId) continue;

  const targetFile = extractCol(colMap.targetFile).replace(/`/g, '').trim();
  
  // Parse dependencies
  const depsRaw = extractCol(colMap.dependencies);
  const dependencies = depsRaw === 'None' || depsRaw === '-' || depsRaw === '' 
    ? [] 
    : depsRaw.replace(/`/g, '').split(/,\s*/).map(d => d.trim()).filter(Boolean);
  
  // Parse arrays safely
  const assignedSkills = parseArray(extractCol(colMap.skills));
  const contextFiles = parseArray(extractCol(colMap.contextFiles));
  const antiGoals = parseArray(extractCol(colMap.antiGoals));
  const description = extractCol(colMap.description);
  
  tasks.push({
    task_id: taskId,
    description,
    dependencies,
    assigned_skills: assignedSkills,
    target_file: targetFile,
    context_files_to_read: contextFiles,
    anti_goals: antiGoals
  });
}

if (tasks.length === 0) {
  console.error(JSON.stringify({ success: false, error: 'No tasks parsed from DAG table' }));
  process.exit(1);
}

// --- Parse API Contracts ---
let apiContracts = {};
const contractSection = masterplan.match(/##\s*🔌\s*API\s*Contracts([\s\S]*?)(?=\n##\s|$)/);
if (contractSection) {
  const jsonBlocks = contractSection[1].match(/```json\n([\s\S]*?)```/g) || [];
  jsonBlocks.forEach(block => {
    try {
      const json = JSON.parse(block.replace(/```json\n?/, '').replace(/```/, ''));
      Object.assign(apiContracts, json);
    } catch (e) { /* skip malformed blocks */ }
  });
}

// --- Generate router.json ---
const now = new Date().toISOString();
const taskStatuses = {};
tasks.forEach(t => { taskStatuses[t.task_id] = 'pending'; });

const router = {
  job_id: jobId,
  project_name: projectName,
  architect_name: architectName,
  manager_name: 'provision-script',
  owner_name: null,
  global_status: 'executing',
  batch_limit: 3,
  created_at: now,
  updated_at: now,
  tasks: taskStatuses
};

fs.writeFileSync(path.join(jobDir, 'router.json'), JSON.stringify(router, null, 2));

// --- Create done/ directory ---
const doneDir = path.join(jobDir, 'done');
if (!fs.existsSync(doneDir)) fs.mkdirSync(doneDir, { recursive: true });

// --- Read Templates ---
const skillRoot = path.resolve(__dirname, '..');
const taskSchemaPath = path.join(skillRoot, 'templates', 'task_schema.json');
const instructionTemplatePath = path.join(skillRoot, 'templates', 'instruction_template.md');

const taskSchemaRaw = fs.existsSync(taskSchemaPath) ? fs.readFileSync(taskSchemaPath, 'utf-8') : null;
const instructionTemplate = fs.existsSync(instructionTemplatePath) ? fs.readFileSync(instructionTemplatePath, 'utf-8') : null;

// --- Provision Each Task ---
const createdFolders = [];

tasks.forEach(task => {
  const taskDir = path.join(jobDir, task.task_id);
  if (!fs.existsSync(taskDir)) fs.mkdirSync(taskDir, { recursive: true });
  
  // Generate task.json
  const taskJson = {
    task_id: task.task_id,
    description: task.description,
    dependencies: task.dependencies,
    assigned_skills: task.assigned_skills,
    target_file: task.target_file,
    current_status: 'pending',
    workers_assigned: [],
    context_files_to_read: task.context_files_to_read,
    anti_goals: task.anti_goals,
    api_contract: apiContracts,
    status_history: [
      {
        status: 'pending',
        timestamp: now,
        changed_by: 'provision-script',
        notes: 'Mapped from masterplan.'
      }
    ]
  };
  
  fs.writeFileSync(path.join(taskDir, 'task.json'), JSON.stringify(taskJson, null, 2));
  
  // Generate instruction.md
  if (instructionTemplate) {
    let instruction = instructionTemplate
      .replace(/\{\{task_id\}\}/g, task.task_id)
      .replace(/\{\{target_file_path\}\}/g, task.target_file)
      .replace(/\{\{assigned_skills_array\}\}/g, JSON.stringify(task.assigned_skills))
      .replace(/\{\{dependencies_list_or_none\}\}/g, task.dependencies.length ? task.dependencies.join(', ') : 'None')
      .replace(/\{\{context_files_to_read_list.*?\}\}/g, task.context_files_to_read.map(f => `* \`${f}\``).join('\n'))
      .replace(/\{\{anti_goals_list.*?\}\}/g, task.anti_goals.map(g => `* ${g}`).join('\n'))
      .replace(/\{\{api_contract_from_task_json\}\}/g, JSON.stringify(apiContracts, null, 2));
    
    // Replace the objective placeholder
    instruction = instruction.replace(
      /\{\{Specific description.*?\}\}/g,
      task.description
    );
    
    fs.writeFileSync(path.join(taskDir, 'instruction.md'), instruction);
  }
  
  createdFolders.push(task.task_id);
});

// --- Output Result ---
console.log(JSON.stringify({
  success: true,
  job_id: jobId,
  tasks_provisioned: createdFolders.length,
  task_ids: createdFolders,
  router_path: path.join(jobDir, 'router.json'),
  done_dir: doneDir
}));

// --- Helpers ---
function parseArray(str) {
  // Handles: ['item1', 'item2'] or `['item1']` or plain text
  const cleaned = str.replace(/`/g, '').trim();
  if (cleaned.startsWith('[')) {
    try {
      return JSON.parse(cleaned.replace(/'/g, '"'));
    } catch (e) {
      return cleaned.replace(/[\[\]']/g, '').split(/,\s*/).map(s => s.trim()).filter(Boolean);
    }
  }
  if (cleaned === 'None' || cleaned === '-' || cleaned === '') return [];
  return cleaned.split(/,\s*/).map(s => s.trim()).filter(Boolean);
}
